src/encode.py

```diff
@@ -3,61 +3,73 @@
-        df = add_feature_counters_for_each_time_window(df, feature, counters, time_windows)
+    for exercise_code in tqdm(exercises, desc=f'Adding the counters for {len(exercises)} exercises'):
+        df = add_exercise_code_counters_for_each_time_window(
+            df, exercise_code, counters, time_windows)
     return df
 
-def add_feature_counters_for_each_time_window(df: pd.DataFrame, feature: str, counters: tuple, time_windows: tuple)-> pd.DataFrame:
+
+def add_exercise_code_counters_for_each_time_window(df: pd.DataFrame, exercise_code: str, counters: tuple, time_windows: tuple) -> pd.DataFrame:
     '''
-    For a given feature, adds a column for all given counters of a student and for all given time windows.
+    For a given exercise_code, adds a column for all given counters of a student and for all given time windows.
     '''
-    for time_window in tqdm(time_windows, desc=f'Adding the counters for the time windows of {feature}'):
-        df = add_feature_counters_in_one_time_window(df, feature, counters, time_window)
+    for time_window in tqdm(time_windows, desc=f'Adding the counters for the time windows of {exercise_code}'):
+        df = add_exercise_code_counters_in_one_time_window(
+            df, exercise_code, counters, time_window)
     return df
 
-def add_feature_counters_in_one_time_window(df: pd.DataFrame, feature: str, counters: tuple, time_window: str)-> pd.DataFrame:
+
+def add_exercise_code_counters_in_one_time_window(df: pd.DataFrame, exercise_code: str, counters: tuple, time_window: str) -> pd.DataFrame:
     '''
-    For a given feature, adds a column for all given counters of a student and for one given time window.
+    For a given exercise_code, adds a column for all given counters of a student and for one given time window.
     '''
     for counter in counters:
-        df = add_one_feature_counter_in_one_time_window(df, feature, counter, time_window)
+        df = add_one_exercise_code_counter_in_one_time_window(
+            df, exercise_code, counter, time_window)
     return df
 
-def add_one_feature_counter_in_one_time_window(df: pd.DataFrame, feature: str, counter: str, time_window: str)-> pd.DataFrame:
-    '''
-    For a given feature, adds a column with the given counter of a student in the specified time window.
-    Possible counters :
-    - wins
-    - attempts
-    '''
+def add_one_exercise_code_counter_in_one_time_window(df: pd.DataFrame, exercise_code: str, counter: str, time_window: str) -> pd.DataFrame:
     df_copy = df.copy()
-    df_copy = df_copy.set_index('timestamp')
-    # rolling works properly on offset only if the timestamp is the index
-    counter_in_the_time_window = df_copy.groupby(by=['student_id', feature], as_index=False).rolling(time_window, closed='left')['correctness']
-    # closed='left' removes the current row from the count and thus prevent data leakage. Think of [-3,0[
+    df_with_only_one_exercise = df[df['exercise_code'] == exercise_code]
```

Comment:
i would call this variable maybe `df_of_exercise` or something that begins with`filtered_df` so we really know that only the rows of exercise_code are included

Reply:
done